python/ray/tests/test_global_state.py

```diff
@@ -108,6 +110,27 @@ def get_state():
     assert get_state() == dead_state
 
 
+def test_global_state_worker_table(ray_start_regular):
+
+    # worker table should be empty at first
+    assert len(ray.state.workers()) == 0
+
+    # add global_worker
+    worker_id = ray.worker.global_worker.worker_id
+    worker_type = gcs_utils.WORKER
+    worker_info = {"key1": "value1"}
+    result = ray.state.state.add_worker(worker_id, worker_type, worker_info)
+
+    # add worker success
+    assert result is True
+
+    # get worker table from gcs
+    workers_data = ray.state.workers()
+
+    assert len(workers_data) == 1
+    assert (binary_to_hex(worker_id) in workers_data) is True
```

Comment:
Can you also test with an actor? Create one more actor and make sure there is an additional worker from the table.

Reply:
yes~

Reply:
hi, @rkooo567 do you mean this?  but it seems to changed nothing.... bacause `GlobalState.add_worker` is the only entry that will write worker info to gcs,  when a raylet started, it's worker will not register info to gcs.

Reply:
No. Ray.workers() is a high level api that is exposed to users, but the test depends ln the implementation details, not the high level behavior. I am not super familiar with this worker table, but I remember it is recorded when a worker failure happened? I’d like this to be tested with actual behavior that creates an entry to a worker table.

Reply:
yes, It indeed, the code segment is `custom_excepthook` method in `worker.py`, but it's a driver type, `GlobalState .workers()` only get the worker type.  or we can test the `get_worker_table()` method in `global_state_accessor` ?  Hmmm.. I think there is no need to test this....

Reply:
Hmm I am not sure if I understood what you are saying. I think my only request is not to use `ray.state.state.add_worker(worker_id, worker_type, worker_info)`, but use more high-level behavior for testing instead. What's the exact issue here?

Reply:
yse, I understand what's you mean, the problem now is: only one place called `GlobalState.add_worker`, It's `custom_excepthook ` method in worker.py,  but it will register a driver type worker to gcs, (`GlobalState .workers()`  only get the worker type worker) except this, no other place will register worker information to gcs(incldue add node, add actor and so on), so, If we don't manually call `ray.state.state.add_worker(worker_id, worker_type, worker_info)`,  the `GlobalState .workers()` will always return empty because no worker will register to gcs.

Reply:
I don't think that's true. For example, look at node_manger.cc

```
void NodeManager::ProcessDisconnectClientMessage(
    const std::shared_ptr<ClientConnection> &client, bool intentional_disconnect) {
  std::shared_ptr<WorkerInterface> worker = worker_pool_.GetRegisteredWorker(client);
  bool is_worker = false, is_driver = false;
  if (worker) {
    // The client is a worker.
    is_worker = true;
  } else {
    worker = worker_pool_.GetRegisteredDriver(client);
    if (worker) {
      // The client is a driver.
      is_driver = true;
    } else {
      RAY_LOG(INFO) << "Ignoring client disconnect because the client has already "
                    << "been disconnected.";
      return;
    }
  }
  RAY_CHECK(!(is_worker && is_driver));
  // If the client has any blocked tasks, mark them as unblocked. In
  // particular, we are no longer waiting for their dependencies.
  if (is_worker && worker->IsDead()) {
    // If the worker was killed by us because the driver exited,
    // treat it as intentionally disconnected.
    intentional_disconnect = true;
    // Don't need to unblock the client if it's a worker and is already dead.
    // Because in this case, its task is already cleaned up.
    RAY_LOG(DEBUG) << "Skip unblocking worker because it's already dead.";
  } else {
    // Clean up any open ray.get calls that the worker made.
    while (!worker->GetBlockedTaskIds().empty()) {
      // NOTE(swang): AsyncResolveObjectsFinish will modify the worker, so it is
      // not safe to pass in the iterator directly.
      const TaskID task_id = *worker->GetBlockedTaskIds().begin();
      AsyncResolveObjectsFinish(client, task_id, true);
    }
    // Clean up any open ray.wait calls that the worker made.
    task_dependency_manager_.UnsubscribeWaitDependencies(worker->WorkerId());
  }

  // Erase any lease metadata.
  leased_workers_.erase(worker->WorkerId());

  // Publish the worker failure.
  auto worker_failure_data_ptr =
      gcs::CreateWorkerFailureData(self_node_id_, worker->WorkerId(), worker->IpAddress(),
                                   worker->Port(), time(nullptr), intentional_disconnect);
  RAY_CHECK_OK(
      gcs_client_->Workers().AsyncReportWorkerFailure(worker_failure_data_ptr, nullptr));
```

Btw, this table name is kind of misleading. This seems to be failed_worker_info or something. But that should be out of topic in this PR.

Reply:
sorry...  @rkooo567 , I did not find a way to simulate this situation in python api, I tried ray.worker.shutdown and then ray.init() to re-init globalWorker, however, the previous workers cannot be obtained yet, maybe you can give me some more clear approach?

Reply:
I will get back to you asap

Reply:
Please look at https://github.com/ray-project/ray/pull/11516#discussion_r511757490